cupy/linalg/_product.py

```diff
@@ -42,6 +42,71 @@
 )
 
 
+def _multi_dot_three(A, B, C, out=None):
+    """Find the best order for three arrays and do the multiplication.
+    """
+    a0, a1b0 = A.shape
+    b1c0, c1 = C.shape
+    # cost1 = cost((AB)C) = a0*a1b0*b1c0 + a0*b1c0*c1
+    cost1 = a0 * b1c0 * (a1b0 + c1)
+    # cost2 = cost(A(BC)) = a1b0*b1c0*c1 + a0*a1b0*c1
+    cost2 = a1b0 * c1 * (a0 + b1c0)
+
+    if cost1 < cost2:
+        return dot(dot(A, B), C, out=out)
+    else:
+        return dot(A, dot(B, C), out=out)
+
+
+def _multi_dot_matrix_chain_order(arr, return_costs=False):
+    """Returns a cupy.ndarray that encodes the optimal order of
+    multiplication. The optimal order array is then used by
+    `_multi_dot()` to do the multiplication.
+
+    It also return the cost matrix if `return_costs` is `True`
+
+    The cost function is given by:
+
+        cost[i, j] = min([
+            cost[prefix] + cost[suffix] + cost_mult(prefix, suffix)
+            for k in range(i, j)])
+    """
+    n = len(arr)
+    # To store dimensions of the matrices
+    p = [a.shape[0] for a in arr] + [arr[-1].shape[1]]
+    # Matrix of costs of the subproblems
+    # m[i,j]: min number of scalar multiplications needed to compute A_{i..j}
+    m = cupy.zeros((n, n), dtype=cupy.double)
+    # Actual ordering
+    # s[i, j] is the value of k at which we split the product A_i..A_j
+    s = cupy.empty((n, n), dtype=cupy.intp)
+
+    for l in range(1, n):
+        for i in range(n - l):
+            j = i + l
+            m[i, j] = cupy.Inf
+            for k in range(i, j):
+                q = m[i, k] + m[k+1, j] + p[i]*p[k+1]*p[j+1]
+                if q < m[i, j]:
```

Comment:
this imposes device/cpu sync.
I think this function should be rewritten into a Cuda kernel using `RawKernel`